# load data for bacteria DE
`B29.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/B29.genes.results")
`B30.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/B30.genes.results")
`B31.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/B31.genes.results")
`B32.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/B32.genes.results")
`LO2.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/LO2.genes.results")
`LO3.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/LO3.genes.results")
`LO4.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/LO4.genes.results")
`LO5.genes` <- read.delim("~/Desktop/Siphamia/siph.genes.results/bacteria/LO5.genes.results")

#make data frames
Genes<- data.frame(B29.genes$gene_id,B29.genes$expected_count,B30.genes$expected_count, B31.genes$expected_count,B32.genes$expected_count, LO2.genes$expected_count,LO3.genes$expected_count, LO4.genes$expected_count,LO5.genes$expected_count,LO5.genes$expected_count)
FPKM<- data.frame(B29.genes$gene_id,B29.genes$FPKM,B30.genes$FPKM,B31.genes$FPKM, B32.genes$FPKM, LO2.genes$FPKM, LO3.genes$FPKM,LO4.genes$FPKM,  LO5.genes$FPKM )

#add column names
names(Genes)<-c("gene_id","B29","B30","B31", "B32","LO2","LO3","LO4", "LO5", "LO5b")

#remove ERCC lines from Genes dataframe
Genes<-Genes[93:4061, ]
Genes<-data.frame(Genes$gene_id,Genes$B29, Genes$B30, Genes$B31, Genes$B32, Genes$LO2, Genes$LO3, Genes$LO4, Genes$LO5, Genes$LO5b)
names(Genes)<-c("gene_id","B29","B30","B31", "B32","LO2","LO3","LO4", "LO5", "LO5b")

#Load edgeR
library("edgeR", lib.loc="~/Desktop/R-3.2.2/library")
#set working directory
setwd("/Users/brisbin/desktop/Siphamia/edgeR_wd")

#read in data
counts <- Genes[ , -c(1,ncol(Genes)) ]
rownames(counts)<-Genes[ , 1] #gene names
colnames(counts)<-paste(c("B29","B30","B31","B32","LO2","LO3","LO4", "LO5"))
View(counts)

##Basic info about the data
#size of data frame
dim(counts)
#column sums - library size
colSums(counts)
#library size in millions
colSums(counts)/1e06
# Number of genes with low counts
table( rowSums( counts ) )[ 1:30 ] 

#convert count matrix to edgeR Object
#create a group variable that tells edgeR which samples belong to which group
group<- c(rep("B",4), rep("LO",4))
cds<-DGEList(counts, group=group)
names(cds)
#original count data
head(cds$counts)
#contains sample information
head(cds$samples)
# How many genes have 0 counts across all samples
sum( cds$all.zeros )


##need to filter out low count reads bc impossible to detect differential expression
#keep only genes with atleast 1 read per million reads in at least 3 samples
#once this is done, can calculate normalisation factors which correct for different compositions of samples
# effective library size = product of actual library size and these factors

cds <- cds[rowSums(1e+06 * cds$counts/expandAsMatrix(cds$samples$lib.size, dim(cds)) > 1) >= 3, ]
dim( cds )
cds <- calcNormFactors( cds )
cds$samples

# effective library sizes
cds$samples$lib.size * cds$samples$norm.factors
#Normalize gene counts to library size --- done?

##from edger docs: an MD plot can show the performance of the TMM normalization, visualizes the library size-adjusted log-fold change between two
#libraries (the difference) against the average log-expression across those libraries (the mean)
#The following MD plot is generated by comparing sample 1 against an artificial library
#constructed from the average of all other samples.
plotMD(cpm(cds, log=TRUE), column=8)
abline(h=0, col="red", lty=2, lwd=2)
#Ideally, the bulk of genes should be centred at a log-fold change of zero. This indicates
#that any composition bias between libraries has been successfully removed. This quality
#check should be repeated by constructing a MD plot for each sample.


#Multi-Dimensional Scaling Plot
#measures similarity of samples and projects measure into 2 dimensions
plotMDS( cds , main = "MDS Plot for Bacteria Count Data", labels = colnames( cds$counts ) )

#Estimating Dispersions
#1st: calculate common dispersion
#each gene get assigned the same dispersion estimate
#output of the estimation includes estimate andother elements added to object cds
cds <- estimateCommonDisp( cds )
names( cds )
#The estimate
cds$common.dispersion


# ##from edger docs
# trt <- factor(c(1,1,1,1,2,2,2,2))
# design <- model.matrix(~trt)
# design <- model.matrix(~group)
# design
# ## only LO, where is B?????
# # there should be another column - B, with 0s and 1's in the opposite positions??
# 
# library("statmod", lib.loc="~/Desktop/R-3.2.2/library")
# cds<- estimateDisp(cds, design, robust=TRUE)
# cds$common.dispersion
# plotBCV(cds)
# fit <- glmQLFit(cds, design, robust=TRUE)
# head(fit$coefficients)
# plotQLDisp(fit)
# #differential expression
# qlf <- glmQLFTest(fit)
# topTags(qlf,n=15)
# #The total number of DE exons in each direction at a FDR of 5% can be examined with
# #decideTestsDGE.
# summary(decideTestsDGE(qlf, p.value=0.05))

#with common dispersion, can estimate tagwise dispersions
#each gene will get its own dispersion estimate
#tagwise dispersions are squeezed towards a common value
#amt of squeezing is governed by parameter prior.n
#higher prior.n --> closer the estimates will be to common dispersion
#default value is nearest interger to 50/(#samples-#groups)
# 50/(6-2)= 12.5 -->13

cds <- estimateTagwiseDisp( cds)
names( cds )
summary( cds$tagwise.dispersion )

#Mean-Variance Plot
#see how well the dispersion factors fit the data
# see raw variance of counts (grey dots), 
#the variance using tagwise dispersions(light blue dots)
#variances using common dispersion (solid blue line)
#variance=mean aka poisson variance (solid black line)

meanVarPlot <- plotMeanVar( cds , show.raw.vars=TRUE ,
                            show.tagwise.vars=TRUE ,
                            show.binned.common.disp.vars=FALSE ,
                            show.ave.raw.vars=FALSE ,
                            dispersion.method = "qcml" , NBline = TRUE ,
                            nbins = 100 ,
                            pch = 16 ,
                            xlab ="Mean Expression (Log10 Scale)" ,
                            ylab = "Variance (Log10 Scale)" ,
                            main = "Bacteria Mean-Variance Plot" )

#Testing
# exactTest() performs pairwise tests for diff exp between 2 groups
# pair indicates which groups should be compared
# output is a list of elements, one of which is a table of results
# de.poi sets dispersion parameter to zero - poisson model - can compare negative binomial results to this

de.cmn <- exactTest( cds , dispersion="common", pair = c( "B" , "LO" ) )
de.tgw <- exactTest( cds , dispersion="tagwise", pair = c( "B" , "LO" ) )
de.poi <- exactTest( cds , dispersion = 1e-06 , pair = c( "B" , "LO" ) )

names( de.tgw )
de.tgw$comparison # which groups have been compared
head( de.tgw$table ) # the results table is in same order of count matrix.

# this does not contain p-values adjusted for multiple testing
# topTags() takes output from exactTest(), adjusts raw p-values using False Discover Rate (FDR) correction
# returns top differentially expressed genes
# sort.by argument allws sorting by p-value, concentration, or fold change
# topTags() returns original counts of the top differentially expressed genes
# set n parameter to number of genes, save the entire topTags() results table

# Top tags for tagwise analysis
options( digits = 3 ) # print only 3 digits
topTags( de.tgw , n = 20 , sort.by = "p.value" ) # top 20 DE genes
# Back to count matrix for tagwise analysis
cds$counts[ rownames( topTags( de.tgw , n = 15 )$table ) , ]
# Sort tagwise results by Fold-Change instead of p-value
resultsByFC.tgw <- topTags( de.tgw , n = nrow( de.tgw$table ) , sort.by = "logFC" )$table
head( resultsByFC.tgw )
# Store full topTags results table
resultsTbl.cmn <- topTags( de.cmn , n = nrow( de.cmn$table ) )$table
resultsTbl.tgw <- topTags( de.tgw , n = nrow( de.tgw$table ) )$table
resultsTbl.poi <- topTags( de.poi , n = nrow( de.poi$table ) )$table
head( resultsTbl.tgw )

#compare p-values to significance level and determine # of diff exp genes
# sig level = ).05
# decideTestsDGE() to determine how many diff exp genes are up or down regulated compared to control

# Names/IDs of DE genes
de.genes.cmn <- rownames( resultsTbl.cmn )[ resultsTbl.cmn$PValue <= 0.05 ]
de.genes.tgw <- rownames( resultsTbl.tgw )[ resultsTbl.tgw$PValue <= 0.05 ]
de.genes.poi <- rownames( resultsTbl.poi )[ resultsTbl.poi$PValue <= 0.05 ]
# Amount significant
length( de.genes.cmn )
length( de.genes.tgw )
length( de.genes.poi )
# Percentage of total genes
length( de.genes.cmn ) / nrow( resultsTbl.cmn ) * 100
length( de.genes.tgw ) / nrow( resultsTbl.tgw ) * 100
length( de.genes.poi ) / nrow( resultsTbl.poi ) * 100
# Up/Down regulated summary for tagwise results
summary( decideTestsDGE( de.tgw , p.value = 0.05 ) ) # the adjusted p-values are used here

# -1 1405
# 0  1181
# 1  1255


#Visualize results
#spread of expression levels for DE genes
#plot concentrations for top 100 DE genes for each analysis

#MA plot shows relationship between concentration and fold-change across genes
# DE genes are red
# non DE genes are black
# organge dots = counts were zero in all samples in one group
# blue line is at log-FC of 2 to represent a level for biological significance
##An MA plot is an application of a Bland???Altman plot for visual representation of two channel DNA microarray gene expression data which has been transformed onto the M (log ratios) and A (mean average) scale.
par( mfrow=c(2,1) )
plotSmear( cds , de.tags=de.genes.poi , main="Bacteria Poisson MA plot" ,
           pair = c("B","LO") ,
           cex = .35 ,
           xlab="Log Concentration" , ylab="Log Fold-Change" )
abline( h = c(-2, 2) , col = "dodgerblue" )
plotSmear( cds , de.tags=de.genes.tgw , main="Bacteria Tagwise MA plot" ,
           pair = c("B","LO") ,
           cex = .35 ,
           xlab="Log Concentration" , ylab="Log Fold-Change" )
abline( h=c(-2,2) , col="dodgerblue" )
par( mfrow=c(1,1) )

# log difference between DE genes under negative binomial/poisson model and tagwise in MA plot
# plot top 500 DE genes, rest are black

par( mfrow = c(2,1) )
plotSmear( cds , de.tags=de.genes.poi[1:500] , main=" Bacteria Poisson MA plot" ,
           pair=c("B","LO") ,
           cex=.5 ,
           xlab="Log Concentration" , ylab="Log Fold-Change" )
abline( h=c(-2,2) , col="dodgerblue" )
plotSmear( cds , de.tags=de.genes.tgw[1:500] , main="Bacteria Tagwise MA plot" ,
           pair=c("B","LO") ,
           cex = .5 ,
           xlab="Log Concentration" , ylab="Log Fold-Change" )
abline( h=c(-2,2) , col="dodgerblue" )
par( mfrow=c(1,1) )

#Output results
#make a table or csv file containing results with concentrations, fold-changes, p-values
#up/down regulated variable, dispersions, and the count matrix

# Change column names to be specific to the analysis, logConc and logFC are the same in both.
colnames( resultsTbl.cmn ) <- c( "logConc" , "logFC" , "pVal.Cmn" , "adj.pVal.Cmn" )
colnames( resultsTbl.tgw ) <- c( "logConc" , "logFC" , "pVal.Tgw" , "adj.pVal.Tgw" )
# Below provides the info to re-order the count matrix to be in line with the order of the results.
wh.rows.tgw <- match( rownames( resultsTbl.tgw ) , rownames( cds$counts ) )
wh.rows.cmn <- match( rownames( resultsTbl.cmn ) , rownames( cds$counts ) )
head( wh.rows.tgw )
# Tagwise Results
combResults.tgw <- cbind( resultsTbl.tgw ,
                          "Tgw.Disp" = cds$tagwise.dispersion[ wh.rows.tgw ] ,
                          "UpDown.Tgw" = decideTestsDGE( de.tgw , p.value = 0.05 )[ wh.rows.tgw ] ,
                          cds$counts[ wh.rows.tgw , ] )
head( combResults.tgw )
# Common Results
combResults.cmn <- cbind( resultsTbl.cmn ,
                          "Cmn.Disp" = cds$common.dispersion ,
                          "UpDown.Cmn" = decideTestsDGE( de.cmn , p.value = 0.05 )[ wh.rows.cmn ] ,
                          cds$counts[ wh.rows.cmn , ] )
head( combResults.cmn )

#combine both common and tagwise results together

wh.rows <- match( rownames( combResults.cmn ) , rownames( combResults.tgw ) )
combResults.all <- cbind( combResults.cmn[,1:4] ,
                          combResults.tgw[wh.rows,3:4] ,
                          "Cmn.Disp" = combResults.cmn[,5],
                          "Tgw.Disp" = combResults.tgw[wh.rows,5],
                          "UpDown.Cmn" = combResults.cmn[,6],
                          "UpDown.Tgw" = combResults.tgw[wh.rows,6],
                          combResults.cmn[,7:ncol(combResults.cmn)] )
head( combResults.all )
# Ouput csv tables of results
write.table( combResults.tgw , file = "combResults_tgw_ex1.csv" , sep = "," , row.names = TRUE )
write.table( combResults.cmn , file = "combResults_cmn_ex1.csv" , sep = "," , row.names = TRUE )
write.table( combResults.all , file = "combResults_all_ex1.csv" , sep = "," , row.names = TRUE )

#par( mfrow=c(3 ,1) )
#hist( resultsTbl.poi[de.genes.poi[1:100],"logConc"] , breaks=10 , xlab="Log Concentration" ,
#      col="red" , xlim=c(-18,-6) , ylim=c(0,0.4) , freq=FALSE , main="Poisson: Top 100" )
#hist( resultsTbl.cmn[de.genes.cmn[1:100],"logConc"] , breaks=25 , xlab="Log Concentration" ,
#      col="green" , xlim=c(-18,-6) , ylim=c(0,0.4) , freq=FALSE , main="Common: Top 100" )
#hist( resultsTbl.tgw[de.genes.tgw[1:100],"logConc"] , breaks=25 , xlab="Log Concentration" ,
#      col="blue" , xlim=c(-18,-6) , ylim=c(0,0.4) , freq=FALSE , main="Tagwise: Top 100" )
#par( mfrow=c(1,1) )